src/lib/Shop.js

```diff
@@ -339,74 +360,119 @@ class AutomationShop
+                    }
+                    else
                     {
-                        return this.__internal__isPokeMarkUnlocked()
-                            || (TownList["Pewter City"].isUnlocked() && Automation.Utils.Route.canMoveToRegion(0))
-                            || TownList["Violet City"].isUnlocked();
-                    }.bind(this);
-            }
-            else if ((item.name == "Dowsing_machine") || (item.name == "Lucky_incense"))
-            {
-                itemData.isUnlocked = function() { return TownList["Lavender Town"].isUnlocked(); };
-                itemData.isPurchasable = function()
+                        sellableItems[item.name].towns.push(townName);
+                    }
+
+                    // Add the requirements
+                    if (shop.requirements !== undefined && shop.requirements.length > 0)
                     {
-                        return this.__internal__isPokeMarkUnlocked()
-                            || (TownList["Lavender Town"].isUnlocked() && Automation.Utils.Route.canMoveToRegion(0))
-                            || TownList["Olivine City"].isUnlocked();
-                    }.bind(this);
-            }
-            else if (item.name == "SmallRestore")
-            {
-                itemData.isUnlocked = function() { return TownList["Cinnabar Island"].isUnlocked(); };
-                itemData.isPurchasable = itemData.isUnlocked;
+                        sellableItems[item.name].requirements.push(shop.requirements);
+                    }
+                    else
+                    {
+                        sellableItems[item.name].requirements.push([]);
+                    }
+                }
+
             }
-            else if (item.name == "MediumRestore")
+        }
+
+        // Iterate with of instead of Object.entries to keep the order
+        for (const itemData of Object.values(sellableItems))
+        {
+            let shopItem = { item: itemData.item }
+            shopItem.isUnlocked = function() {
+                return itemData.towns.some(townName => TownList[townName].isUnlocked())
+                    && (itemData.requirements.length == 0
+                        || itemData.requirements.some(requirementList => requirementList.every(requirement => requirement.isCompleted())));
+            };
+
+            // This may restrict some items that are actually purchasable, but only until the player unlocks the port again
+            shopItem.isPurchasable = () => shopItem.isUnlocked && Automation.Utils.Route.canMoveToRegion(0);
+
+            this.__internal__shopItems.push(shopItem);
+        }
+
+        this.__internal__shopItems.sort(this.__internal__itemSortCompare);
+    }
+
+    static __internal__itemSortCompare(a, b)
+    {
+        // Only bother sorting if both items are the same type
+        if (a.item.currency == b.item.currency)
+        {
+            // Order quest point items by its image type, so all eggs are together
```

Comment:
I'd prefer sorting by class, when possible

Reply:
How, exactly?

This whole sort function is a mess, so any improvements are appreciated.

Reply:
For example : 
  - Eggs : `item instanceof EggItem`
  - Mulch : `item instanceof MulchItem`
  - Shovel : `item instanceof ShovelItem`

Reply:
Still missing this, will work on it later (it is much more involved)